### PR TITLE
[Feature/Extensions] Principal identity for Opensearch requests to/from extensions

### PR DESCRIPTION
**Pull request by DarshitChanpura**
### Description
Adds a basic identifiers to determine owner for each request from OpenSearch to extensions.

### Issues Resolved
* Resolves https://github.com/opensearch-project/opensearch-sdk-java/issues/37

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dbwiddis** [comment]
Hey @DarshitChanpura thanks for this draft PR.   Some quick feedback that should hopefully help you iterate:
 - Don't use the `RegisterRestActionsRequest`.  That's inbound to OpenSearch from an extension that just registers the action on the RestController.
 - Take a look at #4282.  In the [RestActionsRequestHandler](https://github.com/opensearch-project/OpenSearch/pull/4282/files#diff-1aa84083b89a8d6e2b6a2d2c800900607b0ba11df22332342fddf2e4130c1b8f) class we are registering the RestController to respond to a User's REST request by forwarding it to the extension.  This forwarding process is where I think you'll need to insert identity information
   - The [RestSendToExtensionAction](https://github.com/opensearch-project/OpenSearch/pull/4282/files#diff-3c3f553f7f22985d4ce94c12ff1cddc3663f969edf72164c2df106fbaae9fb6b) directly handles the `RestRequest`. You may have to back up and see where that Request comes from and how you can attach a principal to it.  Then we could send it along to the extension with the method and URI, just add a third String parameter which is a key to an internal map where the user is the principal.  (Eventually we'll actually want to send a JWT containing an Access Token.)

**Comment by dbwiddis** [comment]
Continuing on the above comment, the `RestRequest` does include the headers which would have the user's authorization info in it.  So I'd suggest using that class as a basis (it's not merged yet but you can copy it and start with it for now) and:
 - parse the `RestRequest` headers to get the user info (also see #4247 where it looks like @peternied has done this for basic auth) and turn that into a `Principal`
 - store the `Principal` in a map with some unique key.  For now I think a random UUID would work.
 - send the key as a `String` along with the Method and URI in the [RestExecuteOnExtensionRequest](https://github.com/opensearch-project/OpenSearch/pull/4282/files#diff-0751c378a028434526ed265e5faaa70558704942cc29037d17a243d3fd6710c8) class.

Later the extension will send this key back with a different request or transport, and you can look up the `Principal` with it.

**Comment by dbwiddis** [comment]
As far as storing in the map, if you go with Apache Shiro, it looks like the [RememberMeManager](https://shiro.apache.org/static/1.3.1/apidocs/org/apache/shiro/mgt/RememberMeManager.html) is probably built for this purpose.  Not sure if the token it produces is an "access token" but worth looking into.

**Comment by dbwiddis** [comment]
Finally comment for today: where to put the map?  I'd suggest a brand new object somewhere.  Look in Node.java where a lot of individual objects are instantiated, such as the `RestController` and `ExtensionsOrchestrator` and others.  Find a nice home in `org.opensearch.identity` and build a new object to your liking and instantiate it in `Node` and pass it around where it's needed.

File: server/src/main/java/org/opensearch/extensions/RegisterRestActionsRequest.java
Comment (peternied) [inline server/src/main/java/org/opensearch/extensions/RegisterRestActionsRequest.java:27]: I think we would want similar treatment on any request from OpenSearch -> Extension interactions - should a there be a base class for all interactions, or a common wrapper?  @saratvemulapalli have you had thoughts about this space?

Making this more concrete, lets consider the IndicesModuleRequest.

IndicesModuleRequest is translated by ExtensionsOrchestrator to send the request to the Extension (ExtensionRunner).  When the call is send off to the extension, the identity of the operation should be included.  Attaching this information onto of the `Request` object seems most straight forward, but is there another way that would better integrate it?

This would also mean that before transiting the OpenSearch -> Extension boundary there would be some kind of 'get current identity and associate it as a Principle with outbound request' component.

File: server/src/main/java/org/opensearch/identity/Principal.java
Comment (peternied) [inline server/src/main/java/org/opensearch/identity/Principal.java:27]: image

Comment (peternied) [reply server/src/main/java/org/opensearch/identity/Principal.java:27]: I'm not certain the implications of using extending NamedWritable - is there a good reference or implementation that I could follow?

**Review by peternied — CHANGES_REQUESTED** [review]
Thanks for getting this out there and forwarding the discussion around our identities!

File: server/src/main/java/org/opensearch/identity/Principal.java
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/identity/Principal.java:27]: I'm not sure about that yet.. but you can search for NamedWritable in the codebase. It is for arbitrary seralizable objects. See interface definition [here](https://github.com/dbwiddis/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/io/stream/NamedWriteable.java)

File: server/src/main/java/org/opensearch/identity/Principal.java
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/identity/Principal.java:37]: I'm following SCIM definition. However this implementation is open to suggestions and changes

File: server/src/main/java/org/opensearch/identity/Principal.java
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/identity/Principal.java:40]: Refer this comment: https://github.com/opensearch-project/OpenSearch/pull/4028#discussion_r932576648

**Comment by dblock** [comment]
I believe it should be "Username" and "Metadata" vs. "UserName" and "MetaData". 
https://english.stackexchange.com/questions/43436/username-user-name-or-user-name

**Review by peternied — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/identity/ExtensionPrincipalsRegistry.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/identity/ExtensionPrincipalsRegistry.java:23]: If someone gets the identifier from the auth system can they turn around and use that in other requests?

The ideal solution is one single with node that generates these ids, or better yet, an access token.

**Comment by DarshitChanpura** [comment]
will re-base this branch once feature/extensions branch gets rebased after [this PR](https://github.com/opensearch-project/OpenSearch/pull/4328) gets merged.

**Review by peternied — COMMENTED** [review]
I did some brainstorming today with @dbwiddis, attempting to capture that:

What do these identifiers need to do?
* They should be consistent over time (years).
* The identifier should be usable to lookup the principal of an identity.
* If an extension compares two identifiers together, they should be equivalent to one another.
* They should not provide the extension with any additional information.

Potential solutions:
### Use the principal as the identifier
If we control the Idp's, such as the in memory user store, we can made the identifiers effectively random UUID, but we are proxying identity such as through ActiveDirectory, SAML, or OpenId we can be divulging information that the extension shouldn't have - a first/last name or an email address.  If we 'wrap' all Idp's, the solution is the same as below.

### Registration of a random UUID
This requires building a system to vendor and store UUID <-> principal mappings.  Depending on the size of the cluster or the size of the principals this can create scaling limits, support for one million identifiers and/or synchronizing UUID across 1000 node cluster.

### Obfuscate the identifier (current)
Using a transparent encoding process, in this case base64, its really easy to break and we are expecting extension writers not to inspect our open source code to figure out what we are doing.  Seems like we would be providing a false sense of security.

### Encrypt the principal (proposed)
If we used a cryptographic processed, with a key from the OpenSearch service + a key per extension, we can create decryptable principals that are not useful if leaked.  While storage and management of the key inputs are needed, they scale based on the number of extensions.

I know there has been much considerable in this space, but I think this last option will provide the integrity and trade a small amount of computation that prevents scaling overhead.   What do you think of this last proposal?

Note; I think we could naïvely pass the principal through a principal identifier and add the cryptographic process in a follow up PR.

File: server/src/main/java/org/opensearch/extensions/rest/RestSendToExtensionAction.java
Comment (saratvemulapalli) [inline server/src/main/java/org/opensearch/extensions/rest/RestSendToExtensionAction.java:103]: 🙈

File: server/src/main/java/org/opensearch/extensions/rest/RestExecuteOnExtensionRequest.java
Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/extensions/rest/RestExecuteOnExtensionRequest.java:32]: +1 
We ideally do not want 3rdP passing in information to the extension (https://en.wikipedia.org/wiki/Relay_attack)

File: server/src/main/java/org/opensearch/extensions/RegisterRestActionsRequest.java
Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/extensions/RegisterRestActionsRequest.java]: Absolutely, as we are working through the security features I think we should start implementing generic message handler which mandates including node identity, security identity etc. 
May be lets start opening up an issue ?

**Comment by DarshitChanpura** [comment]
> I did some brainstorming today with @dbwiddis, attempting to capture that:
> 
> What do these identifiers need to do?
> 
> * They should be consistent over time (years).
> * The identifier should be usable to lookup the principal of an identity.
> * If an extension compares two identifiers together, they should be equivalent to one another.
> * They should not provide the extension with any additional information.
> 
> Potential solutions:
> 
> ### Use the principal as the identifier
> If we control the Idp's, such as the in memory user store, we can made the identifiers effectively random UUID, but we are proxying identity such as through ActiveDirectory, SAML, or OpenId we can be divulging information that the extension shouldn't have - a first/last name or an email address. If we 'wrap' all Idp's, the solution is the same as below.
> 
> ### Registration of a random UUID
> This requires building a system to vendor and store UUID <-> principal mappings. Depending on the size of the cluster or the size of the principals this can create scaling limits, support for one million identifiers and/or synchronizing UUID across 1000 node cluster.
> 
> ### Obfuscate the identifier (current)
> Using a transparent encoding process, in this case base64, its really easy to break and we are expecting extension writers not to inspect our open source code to figure out what we are doing. Seems like we would be providing a false sense of security.
> 
> ### Encrypt the principal (proposed)
> If we used a cryptographic processed, with a key from the OpenSearch service + a key per extension, we can create decryptable principals that are not useful if leaked. While storage and management of the key inputs are needed, they scale based on the number of extensions.
> 
> I know there has been much considerable in this space, but I think this last option will provide the integrity and trade a small amount of computation that prevents scaling overhead. What do you think of this last proposal?
> 
> Note; I think we could naïvely pass the principal through a principal identifier and add the cryptographic process in a follow up PR.

I completely agree with the proposed solution of sending encrypted information to extensions. However for this initial PR are we to pass the name string as is or something other than this should be passed?

**Comment by saratvemulapalli** [comment]
> I completely agree with the proposed solution of sending encrypted information to extensions. However for this initial PR are we to pass the name string as is or something other than this should be passed?

I like the proposal. @DarshitChanpura lets make it iterative (a.k.a be scrappy). We can start with a String for now and lets get the initial change across OpenSearch <-> Extension working and then we can start working on encryption.

File: server/src/main/java/org/opensearch/identity/ExtensionIdentifier.java
Comment (DarshitChanpura) [inline server/src/main/java/org/opensearch/identity/ExtensionIdentifier.java:23]: to better name it, let's go with `Principal`

File: server/src/main/java/org/opensearch/extensions/rest/RestExecuteOnExtensionRequest.java
Comment (peternied) [inline server/src/main/java/org/opensearch/extensions/rest/RestExecuteOnExtensionRequest.java:66]: While the token's type is very illustrative of what it is, `token` doesn't convey much meaning, how about getRequestIssuerIdentity()?

**Review by peternied — COMMENTED** [review]
Looking really close

File: server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java:39]: Ideally there should be but I'm not sure whether it will actually be used. I will add it for now.

File: server/src/main/java/org/opensearch/extensions/rest/RestExecuteOnExtensionRequest.java
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/extensions/rest/RestExecuteOnExtensionRequest.java:66]: I will leave it to getToken here unless we want to rename the field to `requestIssuerIdentity`, but i will rename `generateToken` to `generateRequestIssuerIdentity`. Thoughts?

File: CHANGELOG.md
Comment (dbwiddis) [inline CHANGELOG.md:19]: FYI, see #4477 and #4479.  This may be moved to a separate section in the future. Your thoughts on that issue and PR welcome!

File: server/src/main/java/org/opensearch/identity/package-info.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/package-info.java:12]: Package info shouldn't reference a specific class, but the type of classes in this package.

File: server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java:23]: I don't think there's any value in implementing Serializable, and I've read multiple places it should not be used in new code, and only used when required to interact with legacy code.

Its removal from Java is a [long term goal](https://adtmag.com/articles/2018/05/30/java-serialization.aspx).

File: server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java:27]: Please add a javadoc here explaining what type of String is expected.  Is this a JWT?  Any String?  I see elsewhere in `extractToken()` it is expected to have two fields around a colon, but don't JWTs have two colons?  

Consider throwing an `IllegalArgumentException` for malformed tokens here, so you don't get the exception on `getToken()` in other code.

File: server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java:45]: It will be used in your test class `PrincipalIdentifierTokenTest` that I'm certain you're about to write.

File: server/src/test/java/org/opensearch/extensions/rest/RestExecuteOnExtensionTests.java
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/extensions/rest/RestExecuteOnExtensionTests.java:31]: This doesn't conform to the format expected by `extractToken()`.  (Which indicates that method is not tested...)

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:31]: First sentence of a javadoc should be more descriptive.

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:37]: The javadoc says it's encrypted but this result isn't.  Add a comment (and Issue matching the comment) to add encryption later.   

Should we be generating a JWT here?

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:50]: While this matches the current placeholder implementation this isn't the final decryption. Need a lot more comments here explaining the future plan for encrypt/decrypt here.

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:55]: Does this need `@Override` annotation?  Will it still have access to `parts` when this method finishes executing?  I'm not confident this will work as intended (a test showing that it does would help)

File: server/src/main/java/org/opensearch/extensions/rest/RestSendToExtensionAction.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/rest/RestSendToExtensionAction.java:55]: This needs a test.

File: server/src/test/java/org/opensearch/extensions/rest/RestExecuteOnExtensionTests.java
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/extensions/rest/RestExecuteOnExtensionTests.java:37]: `PrincipalIdentifierToken` doesn't have a no-arg constructor, does this code work?

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:20]: if this is only set by the constructor it should be final. If it's not intended to be final, add a setter.

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:45]: ... or null, but shouldn't it throw IllegalArgumentException rather than return null?

Should you also verify that the uniqueID matches for this particular processor?

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Good progress, enough to see what you're trying to do.  Needs a lot more documentation and tests.

**Comment by cwperks** [comment]
I was able to get by the error by adding `PrincipalIdentifierToken` to the registry in `NetworkModule`. https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/network/NetworkModule.java#L128

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (peternied) [reply server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:50]: @DarshitChanpura Could you create issue(s) and then link them in here?

File: CHANGELOG.md
Comment (DarshitChanpura) [reply CHANGELOG.md:19]: Should we have a section that keeps changelog in Extensions domain, so it doesn't mix up with changes in core.

File: server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java:23]: was originally on the fence of keeping, but will remove it!

File: server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java:27]: For now the input is a simple colon-separated String, but once Encryption/Decryption is implemented we can update the documentation here. +1 to throwing exceptions part.

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:37]: I'm not sure about that yet, but we can cover that as part of adding encryption/decryption task.
I will update the javadoc

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:45]: yes it should be verified, but I think it should be implemented with encryption/decryption task

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:50]: Sure. [The issue](https://github.com/opensearch-project/OpenSearch/issues/4485) is already created. I will add them here.

File: server/src/test/java/org/opensearch/extensions/rest/RestExecuteOnExtensionTests.java
Comment (peternied) [reply server/src/test/java/org/opensearch/extensions/rest/RestExecuteOnExtensionTests.java:37]: The test was executed locally and passed.  To registry an entry it needs a `(InputStream) -> T` function, see [Writeable](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/io/stream/Writeable.java#L94) - we got the opportunity to learn all about NamedWriteables when authoring this code :D

File: server/src/test/java/org/opensearch/extensions/rest/RestExecuteOnExtensionTests.java
Comment (peternied) [reply server/src/test/java/org/opensearch/extensions/rest/RestExecuteOnExtensionTests.java:31]: This identifier stores the token object (a string), and its wire-safe since it doesn't include any of identity information (principal/subject).

IMO this is correct for this file; however for `RestSendToExtensionAction.java` something should have broken.

Comment (DarshitChanpura) [reply server/src/test/java/org/opensearch/extensions/rest/RestExecuteOnExtensionTests.java:31]: A separate check should be implemented to assert the validity of the token. However, I'm unsure about whether it should be implemented now or once encryption/decryption is in place

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:55]: I'm unsure whether it would need an Override annotation, conventionally it should. However the tests passed with/without the annotation.

File: server/src/test/java/org/opensearch/extensions/rest/RestExecuteOnExtensionTests.java
Comment (DarshitChanpura) [reply server/src/test/java/org/opensearch/extensions/rest/RestExecuteOnExtensionTests.java:31]: I have implemented a simple check that checks validity of token based on current input format

Comment (DarshitChanpura) [reply server/src/test/java/org/opensearch/extensions/rest/RestExecuteOnExtensionTests.java:31]: I have changed the way the PrincipalIdentifierToken is instantiated in this class as we don't expected it to be instantiated with a raw string ever. Instead it should always call extensionTokenProcessor.generateToken(principal) to return an object.

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:55]: I have changed it to a lambda. `() -> parts[0];`

Comment (peternied) [reply server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:55]: Java might do the right thing, but the clearest way to resolve this is assign the value to a final local variable that is passed to the class.
```
final String name = parts[0];
return new Principal() {
    public String getName() {
        return name;
    }
}
```

**Comment by DarshitChanpura** [comment]
Pre-commit [failed with ](https://github.com/opensearch-project/OpenSearch/actions/runs/3055720723/jobs/4929095222#step:4:75) unable to resolve netty dependencies. Unable to reproduce it locally.

File: server/src/main/java/org/opensearch/extensions/rest/RestSendToExtensionAction.java
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/extensions/rest/RestSendToExtensionAction.java:55]: This should be tested inside the test for `prepareRequest` method. Is there a place where this test is already written?

**Comment by DarshitChanpura** [comment]
Build failed with this:
```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':distribution:bwc:staged:buildBwcLinuxTar'.
> Building 2.3.0 didn't generate expected file /var/jenkins/workspace/gradle-check/search/distribution/bwc/staged/build/bwc/checkout-2.3/distribution/archives/linux-tar/build/distributions/opensearch-min-2.3.0-SNAPSHOT-linux-x64.tar.gz
```
Probably need to re-trigger the check

File: CHANGELOG.md
Comment (dbwiddis) [reply CHANGELOG.md:22]: Yeah now that #4477 is merged you should add to that section.  Won't break anything but will add extra work when main releases new version or when we finally merge.

**Review by dbwiddis — APPROVED** [review]

File: server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java
Comment (saratvemulapalli) [inline server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java:20]: This is very much tied to an extension, do you want to do something more generic which we can use for non-extensions? 

Again, not a blocker for this PR. Trying to understand how we could make it better for rest of the use-cases.

**Review by saratvemulapalli — APPROVED** [review]
Overall the changes look good to me. @DarshitChanpura thanks for this start.

Dropped a couple of questions but it shouldn't be blocking the merge.

File: server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java:22]: How about we create a new javadoc flag and call it as `opensearch.extensions`?

Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java:22]: As we refactor, I think we would want to call out for both plugins and extensions.

Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/identity/PrincipalIdentifierToken.java:22]: I'll create a follow up.

**Pull request merged**
